### ISSUE TITLE
Replica restart cannot enter normal synchronization state

### ISSUE THREAD

**Issue by boomballa** [dragonflydb/dragonfly#1796]
**Describe the bug**
The slave library executes shutdown, then restarts, synchronizes the thread, and attempts to re-establish the master-slave relationship, but ultimately fails to report an error.

```
# Replication
role:replica
master_host: dfly-master-ip
master_port:6379
**master_link_status:down**
master_last_io_seconds_ago:11407
master_sync_in_progress:0
```

The information in the ERROR log has a hint:

```
I20230904 11:29:59.056198 1972224 init.cc:69] dragonfly running in opt mode.
I20230904 11:29:59.056485 1972224 dfly_main.cc:666] Starting dragonfly df-v1.8.0-7c99d2d1111e2636556ad8c2afad43c396906d01
I20230904 11:29:59.056677 1972224 dfly_main.cc:722] Max memory limit is: 196.00GiB
I20230904 11:29:59.063757 1972224 proactor_pool.cc:147] Running 56 io threads
I20230904 11:29:59.064626 1972224 dfly_main.cc:397] Listening on admin socket dfly-master-ip:16379
I20230904 11:29:59.135763 1972224 server_family.cc:171] Data directory is "/dragonfly/dragonfly6379"
I20230904 11:29:59.136832 1972224 server_family.cc:698] Loading /dragonfly/dragonfly6379/dump-2023-09-04T11:29:29-summary.dfs
I20230904 11:29:59.149757 1972226 listener_interface.cc:97] sock[116] AcceptServer - listening on port 16379
I20230904 11:29:59.149732 1972230 listener_interface.cc:97] sock[173] AcceptServer - listening on port 6379
E20230904 11:30:19.243443 1972225 server_family.cc:1488] Subcommand KILL not supported
E20230904 11:30:19.243538 1972225 server_family.cc:1488] Subcommand KILL not supported
E20230904 11:30:19.306025 1972227 server_family.cc:1488] Subcommand KILL not supported
E20230904 11:30:19.306118 1972227 server_family.cc:1488] Subcommand KILL not supported
I20230904 11:31:41.181810 1972228 server_family.cc:743] Load finished, num keys read: 175200434
I20230904 11:31:49.736783 1972227 server_family.cc:2008] Replicating dfly-replica-ip:6379
I20230904 11:31:51.995817 1972227 protocol_client.cc:219] Resetting endpoint! dfly-replica-ip, 6379
I20230904 11:31:52.065325 1972227 server_family.cc:2008] Replicating dfly-replica-ip:6379
W20230904 11:32:08.474423 1972231 **dfly_main.cc:116] Remote version - connection error [version.dragonflydb.io:443] : Connection timed out**
```
<img width="926" alt="image" src="https://github.com/dragonflydb/dragonfly/assets/8543659/e5e79b0c-f541-4d5e-bbef-1d413b86d16f">

After the instance is restarted, it will indeed perform full synchronization directly with the main database, but it seems that due to some reason, the full synchronization failed, it did not enter the `stable sync` state, and then cleared its own data again.

<img width="682" alt="image" src="https://github.com/dragonflydb/dragonfly/assets/8543659/2870431e-79b2-47ae-b11f-41346c0911a7">

When I execute slaveof no one or replicaof no one, it will be blocked, and I observe that the instance does not perform operations such as full pull again. In the log, I can only see that the command has indeed been executed more than once.

```
# Replication
role:replica
master_host:dfly-master-ip
master_port:6379
**master_link_status:up**
master_last_io_seconds_ago:0
master_sync_in_progress:0
```

**Additional context**
When I cleared the replica instance data directory and restarted, it still performed full sync from the master again, and then entered the `stable sync` stage, and it can work normally.

I have two questions here:
1. Where is the replicaof information stored in dfly? Because I observed that there is no configuration similar to slaveof or replicaof in dfly.conf. Why can I still contact the master directly and restart full and incremental data synchronization after clearing the data directory and restarting?
2. After the replica instance is shut down normally and the data directory is not cleaned, why does it become blocked? What is causing this and what should I do correctly when the instance is in `master_link_status:down` state?

**Comment by adiholden** [comment]
Hi @boomballa 
Thanks for contacting and creating the issue.
Can you share how you configured replication? did you connect redis-cli to the replica and run replicaof command? 
Can you share the configuration commands that you run on server replica and master, so I will have a better understating of you flow? how do you restart replica? is your master and replica version is 1.8?
1. When you use replicaof command via cli this information is not stored to disk, this is just a state that server enters.
 If you are using a config file to run replicaof there is no other place the server saves this information.
2. master_link_status_down means that the replica was not able to connect to master. running replicaof no one should never block and we need to investigate why this happens. I will be able to check more into this after I get will the reply for the questions above.

**Comment by boomballa** [comment]
> Hi @boomballa Thanks for contacting and creating the issue. Can you share how you configured replication? did you connect redis-cli to the replica and run replicaof command? Can you share the configuration commands that you run on server replica and master, so I will have a better understating of you flow? how do you restart replica? is your master and replica version is 1.8?
> 
> 1. When you use replicaof command via cli this information is not stored to disk, this is just a state that server enters.
>    If you are using a config file to run replicaof there is no other place the server saves this information.
> 2. master_link_status_down means that the replica was not able to connect to master. running replicaof no one should never block and we need to investigate why this happens. I will be able to check more into this after I get will the reply for the questions above.

@adiholden Hello, thanks for the reply.
1. The first point you raised is also the point that I am more confused about. I did not configure the replicaof information in the dfly configuration file. When the replica is restarted, I understand that the initial master-slave state should be an independent master. When I After re-executing replicaof, establish a replication relationship with the master I specified.

But currently, it seems that after the slave library is restarted, it will automatically initiate a replication request with the previous master. I have tested that this situation can be reproduced.

So I am very confused, when the replica instance machine does not store replication information, why is the previous replication relationship automatically established after the instance is restarted? Is it because the previous replica information is stored in the master's dictionary? When the replica is detected to be restarted After that, force the connection to be reestablished?
I feel like I need your help confirming the logic here.

2.  The versions of the master and replica deployed here are both v1.8. The first time the master-slave replication relationship is established is by executing the command `repilcaof master_ip master_port`  on the slave library, and everything is normal at first.
Then I changed the configuration of a replica instance, then used the shutdown command to shut down the instance, and then used the startup command `nohup dragonfly --flagfile=/dragonfly/dragonfly6379/dragonfly6379.conf &` to restart the instance, and then caused the above problems.
To be sure, when the info information shows `master_link_status down`, my master status is healthy and the network level is completely OK when I use telnet to test. At the same time, the business can also use the service normally, so I judge that it should This problem is not caused by the inability of the master to connect, but after the full sync is completed and the attempt to enter the stable state fails, which leads to the generation of the master_link_status down state, and there is no attempt to re-establish the connection again, and the state has been maintained like this.
When I tried the `replicaof master_ip master_port `command again, it was blocked. I observed that full sync was not performed again, and I could only force quit the client.
In the end, I cleared all the bgsave files in dfly's data directory and restarted the instance before it returned to normal again.
